File: julia/src/data_structures/hash_table.jl
"""
Hash Table - Open addressing with linear probing

Time Complexity (average):
- put!: O(1)
- get: O(1)
- remove!: O(1)
- contains: O(1)
"""

const EMPTY = 0
const OCCUPIED = 1
const DELETED = 2

mutable struct HTEntry{K,V}
    key::K
    value::V
    state::Int

    HTEntry{K,V}() where {K,V} = new{K,V}()
    HTEntry{K,V}(k::K, v::V, s::Int) where {K,V} = new{K,V}(k, v, s)
end

mutable struct HashTable{K,V}
    buckets::Vector{HTEntry{K,V}}
    count::Int
    capacity::Int

    function HashTable{K,V}(capacity::Int=16) where {K,V}
        buckets = [HTEntry{K,V}() for _ in 1:capacity]
        for b in buckets
            b.state = EMPTY
        end
        new{K,V}(buckets, 0, capacity)
    end
end

HashTable() = HashTable{Any,Any}()

function _hash_index(ht::HashTable{K,V}, key::K) where {K,V}
    (hash(key) % ht.capacity) + 1
end

function _find_slot(ht::HashTable{K,V}, key::K) where {K,V}
    idx = _hash_index(ht, key)
    first_deleted = 0

    for _ in 1:ht.capacity
        entry = ht.buckets[idx]
        if entry.state == EMPTY
            return first_deleted != 0 ? first_deleted : idx
        elseif entry.state == DELETED
            if first_deleted == 0
                first_deleted = idx
            end
        elseif entry.key == key
            return idx
        end
        idx = (idx % ht.capacity) + 1
    end

    first_deleted != 0 ? first_deleted : 0
end

function _resize!(ht::HashTable{K,V}) where {K,V}
    old_buckets = ht.buckets
    ht.capacity *= 2
    ht.buckets = [HTEntry{K,V}() for _ in 1:ht.capacity]
    for b in ht.buckets
        b.state = EMPTY
    end
    ht.count = 0

    for entry in old_buckets
        if entry.state == OCCUPIED
            put!(ht, entry.key, entry.value)
        end
    end
end

"""
    put!(ht::HashTable{K,V}, key::K, value::V) where {K,V}

Insert or update a key-value pair.
"""
function put!(ht::HashTable{K,V}, key::K, value::V) where {K,V}
    if (ht.count + 1) / ht.capacity > 0.75
        _resize!(ht)
    end

    idx = _find_slot(ht, key)
    if idx == 0
        _resize!(ht)
        idx = _find_slot(ht, key)
    end

    if ht.buckets[idx].state != OCCUPIED
        ht.count += 1
    end
    ht.buckets[idx] = HTEntry{K,V}(key, value, OCCUPIED)
    ht
end

Base.setindex!(ht::HashTable{K,V}, v::V, k::K) where {K,V} = put!(ht, k, v)

"""
    get(ht::HashTable{K,V}, key::K) where {K,V}

Get value by key. Throws KeyError if not found.
"""
function get(ht::HashTable{K,V}, key::K) where {K,V}
    idx = _hash_index(ht, key)

    for _ in 1:ht.capacity
        entry = ht.buckets[idx]
        if entry.state == EMPTY
            throw(KeyError(key))
        elseif entry.state == OCCUPIED && entry.key == key
            return entry.value
        end
        idx = (idx % ht.capacity) + 1
    end

    throw(KeyError(key))
end

Base.getindex(ht::HashTable, key) = get(ht, key)

"""
    get(ht::HashTable{K,V}, key::K, default::V) where {K,V}

Get value by key, or return default if not found.
"""
function get(ht::HashTable{K,V}, key::K, default::V) where {K,V}
    try
        return get(ht, key)
    catch e
        if isa(e, KeyError)
            return default
        end
        rethrow(e)
    end
end

"""
    contains(ht::HashTable{K,V}, key::K) where {K,V}

Check if key exists.
"""
function contains(ht::HashTable{K,V}, key::K) where {K,V}
    idx = _hash_index(ht, key)

    for _ in 1:ht.capacity
        entry = ht.buckets[idx]
        if entry.state == EMPTY
            return false
        elseif entry.state == OCCUPIED && entry.key == key
            return true
        end
        idx = (idx % ht.capacity) + 1
    end

    false
end

Base.haskey(ht::HashTable, key) = contains(ht, key)

"""
    remove!(ht::HashTable{K,V}, key::K) where {K,V}

Remove a key-value pair. Returns true if key was found.
"""
function remove!(ht::HashTable{K,V}, key::K) where {K,V}
    idx = _hash_index(ht, key)

    for _ in 1:ht.capacity
        entry = ht.buckets[idx]
        if entry.state == EMPTY
            return false
        elseif entry.state == OCCUPIED && entry.key == key
            ht.buckets[idx].state = DELETED
            ht.count -= 1
            return true
        end
        idx = (idx % ht.capacity) + 1
    end

    false
end

Base.delete!(ht::HashTable, key) = remove!(ht, key)

"""
    isempty(ht::HashTable)

Check if the hash table is empty.
"""
Base.isempty(ht::HashTable) = ht.count == 0

"""
    length(ht::HashTable)

Return the number of entries.
"""
Base.length(ht::HashTable) = ht.count

"""
    size(ht::HashTable)

Return the number of entries.
"""
Base.size(ht::HashTable) = ht.count

"""
    empty!(ht::HashTable)

Remove all entries.
"""
function Base.empty!(ht::HashTable{K,V}) where {K,V}
    for i in 1:ht.capacity
        ht.buckets[i].state = EMPTY
    end
    ht.count = 0
    ht
end

"""
    keys(ht::HashTable{K,V}) where {K,V}

Get all keys.
"""
function Base.keys(ht::HashTable{K,V}) where {K,V}
    result = K[]
    for entry in ht.buckets
        if entry.state == OCCUPIED
            push!(result, entry.key)
        end
    end
    result
end

"""
    values(ht::HashTable{K,V}) where {K,V}

Get all values.
"""
function Base.values(ht::HashTable{K,V}) where {K,V}
    result = V[]
    for entry in ht.buckets
        if entry.state == OCCUPIED
            push!(result, entry.value)
        end
    end
    result
end

# Iterator interface
function Base.iterate(ht::HashTable, state=1)
    while state <= ht.capacity
        if ht.buckets[state].state == OCCUPIED
            entry = ht.buckets[state]
            return ((entry.key, entry.value), state + 1)
        end
        state += 1
    end
    nothing
end

File: lua/data_structures/hash_table.lua
-- Hash Table - Open addressing with linear probing
--
-- Time Complexity (average):
-- - put: O(1)
-- - get: O(1)
-- - remove: O(1)
-- - contains: O(1)

local HashTable = {}
HashTable.__index = HashTable

local EMPTY = 0
local OCCUPIED = 1
local DELETED = 2

local function hashString(s)
    local h = 0
    for i = 1, #s do
        h = (h * 31 + string.byte(s, i)) % 2147483647
    end
    return h
end

local function hashValue(v)
    local t = type(v)
    if t == "string" then
        return hashString(v)
    elseif t == "number" then
        return math.floor(v) % 2147483647
    else
        return hashString(tostring(v))
    end
end

function HashTable.new(capacity)
    local self = setmetatable({}, HashTable)
    self.capacity = capacity or 16
    self.buckets = {}
    self.count = 0
    for i = 1, self.capacity do
        self.buckets[i] = { state = EMPTY }
    end
    return self
end

function HashTable:_hashIndex(key)
    return (hashValue(key) % self.capacity) + 1
end

function HashTable:_findSlot(key)
    local index = self:_hashIndex(key)
    local firstDeleted = nil

    for _ = 1, self.capacity do
        local entry = self.buckets[index]
        if entry.state == EMPTY then
            return firstDeleted or index
        elseif entry.state == DELETED then
            if firstDeleted == nil then
                firstDeleted = index
            end
        elseif entry.key == key then
            return index
        end
        index = (index % self.capacity) + 1
    end

    return firstDeleted or 0
end

function HashTable:_resize()
    local oldBuckets = self.buckets
    local oldCapacity = self.capacity
    self.capacity = self.capacity * 2
    self.buckets = {}
    self.count = 0
    for i = 1, self.capacity do
        self.buckets[i] = { state = EMPTY }
    end
    for i = 1, oldCapacity do
        if oldBuckets[i].state == OCCUPIED then
            self:put(oldBuckets[i].key, oldBuckets[i].value)
        end
    end
end

function HashTable:put(key, value)
    if (self.count + 1) / self.capacity > 0.75 then
        self:_resize()
    end
    local index = self:_findSlot(key)
    if self.buckets[index].state ~= OCCUPIED then
        self.count = self.count + 1
    end
    self.buckets[index] = { key = key, value = value, state = OCCUPIED }
end

function HashTable:get(key)
    local index = self:_hashIndex(key)
    for _ = 1, self.capacity do
        local entry = self.buckets[index]
        if entry.state == EMPTY then
            return nil
        elseif entry.state == OCCUPIED and entry.key == key then
            return entry.value
        end
        index = (index % self.capacity) + 1
    end
    return nil
end

function HashTable:contains(key)
    return self:get(key) ~= nil
end

function HashTable:remove(key)
    local index = self:_hashIndex(key)
    for _ = 1, self.capacity do
        local entry = self.buckets[index]
        if entry.state == EMPTY then
            return false
        elseif entry.state == OCCUPIED and entry.key == key then
            self.buckets[index].state = DELETED
            self.count = self.count - 1
            return true
        end
        index = (index % self.capacity) + 1
    end
    return false
end

function HashTable:isEmpty()
    return self.count == 0
end

function HashTable:size()
    return self.count
end

function HashTable:clear()
    self.buckets = {}
    self.count = 0
    for i = 1, self.capacity do
        self.buckets[i] = { state = EMPTY }
    end
end

function HashTable:keys()
    local result = {}
    for i = 1, self.capacity do
        if self.buckets[i].state == OCCUPIED then
            result[#result + 1] = self.buckets[i].key
        end
    end
    return result
end

function HashTable:values()
    local result = {}
    for i = 1, self.capacity do
        if self.buckets[i].state == OCCUPIED then
            result[#result + 1] = self.buckets[i].value
        end
    end
    return result
end

return HashTable
